R/utils.R

```diff
@@ -299,7 +299,34 @@ create_app_id <- function(data, modules) {
   data <- if (inherits(data, "teal_data")) {
     as.list(data@env)
   } else if (inherits(data, "teal_data_module")) {
-    body(data$server)
+    deparse1(body(data$server))
   }
+  modules <- lapply(flatten_modules(modules), function(x) lapply(x, defunction))
+
   rlang::hash(list(data = data, modules = modules))
 }
+
+#' Flatten `teal_modules` to a list of `teal_module`s.
+#' @keywords internal
+#' @noRd
+flatten_modules <- function(modules) {
+  if (inherits(modules, "teal_module")) {
+    list(modules)
```

Comment:
This will never be reached due to `checkmate::assert_class(modules, "teal_modules")` above

Reply:
`teal_modules` can have `teal_modules` elements as well as `teal_module`.

Reply:
Ah you're right. I didn't see the recursion here.

Why visit all elements twice though?

Reply:
Unless the idea is to make the App ID resilient to name changes, it seems the same can be achieved by visiting all the elements just once.

The `defunction` will visit all `teal_module` and `teal_modules`

```r
> is.list(modules(example_module()))
[1] TRUE
> is.list(example_module())
[1] TRUE
> 
```

```r
mods <- modules(
  label = "tab1",
  modules(
    modules(
      teal::example_module("example1", datanames = NULL),
      modules(
        modules(
          teal::example_module("example1", datanames = NULL)
        )
      )
    )
  ),
  teal::example_module("example2", datanames = "iris")
)

defunction(mods)
```

Reply:
You're right, `defunction` already handles a nested list.
Removing.